piedpiper/gman/client.py

```diff
@@ -0,0 +1,255 @@
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get tasks for thread_id {thread_id}. \n\n{e}"
+        )
+    else:
+        if query_filter:
+            return list(filter(query_filter, r.json()))
+        else:
+            return r.json()
+
+
+def get_thread_id_events(thread_id=None, gman_url=None, query_filter=None):
+    """
+    Get list of all events for a given thread_id, optionally filtered by
+    a lambda expression.
+    :param thread_id: Thread ID to query for as a string.
+    :param gman_url:  GMan endpoint as a string.
+    :param query_filter: Lambda expression
+    :return: List of task events
+    """
+    try:
+        r = requests.get(f"{gman_url}/thread/{thread_id}/events")
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        raise requests.exceptions.HTTPError(
+            f"Gman returned with a bad status code. \n\n{e}"
+        )
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get tasks for thread_id {thread_id}. \n\n{e}"
+        )
+    else:
+        if query_filter:
+            return list(filter(query_filter, r.json()))
+        else:
+            return r.json()
+
+
+def wait_for_thread_id_complete(thread_id=None, gman_url=None, retry_max=10):
+    """
+    Wait for all tasks under a given thread_id to return with a status of complete,
+    up to the retry_max.
+    :param thread_id: thread_id as a string to search for
+    :param gman_url: GMan endpoint as a string
+    :param retry_max: Number of times to retry as an integer
+    :return: True or exception
+    """
+    retries = 0
+    while retries < retry_max:
+        try:
+            log.debug(f"Checking status of thread {thread_id}")
+            r = requests.head(f"{gman_url}/thread/{thread_id}")
+            r.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            raise requests.exceptions.HTTPError(
+                f"GMan returned with a bad status code. \n\n{e}"
+            )
+        except requests.exceptions.RequestException as e:
+            raise requests.exceptions.HTTPError(
+                f"Failed to check status of task. \n\n{e}"
+            )
+        else:
+            running, completed, failed = [
+                r.headers.get(key)
+                for key in [
+                    "x-gman-tasks-running",
+                    "x-gman-tasks-completed",
+                    "x-gman-tasks-failed",
+                ]
+            ]
+            if int(failed) > 0:
+                raise TaskError(f"Thread {thread_id} has failures")
+            elif int(running) > 0:
+                retries += 1
+                time.sleep(1)
+            elif int(running) == 0 and int(completed) > 0:
+                return True
+
+    raise TimeoutError(f"Checking thread_id status timed out for thread_id {thread_id}")
```

Comment:
TimeoutError, could this be named ExecTimoutError... also where is this coming from? I don't see it imported or defined

Reply:
or maybe ThreadTimeoutError... or DeadThreadError ;)

Reply:
It's a python builtin! We could make another exception but I opted to use one that aptly described the error without making my own.

Reply:
oh so it is, cool yeah no need to go out and make a new one.